### PR TITLE
Permanent default variables

### PR DESCRIPTION
**Pull request by MathiasPaumgarten**
This is first proposed by @nabto in #92 

**TL;DR** The idea is that every bake task has fixed permanent variables that are filled by the system. The way node has `__dirname` for instance. 

Currently a little bit of a work in progress. I haven't nailed down the exact variables that are reasonable. For now we have `__filename` and `__timestamp`. That is a little misleading already because `__filename` is really `__filepath`.

This is a in-progress branch. I just wanted to keep it separate from master so @david-zacharias can have a look at it and give me his opinion (since you built most of the features lately)

Some TODOs:
- cleaner more extensive tests
- possibly broader/different set of permanent variables
- maybe cleaner implementation, though I'm not sure how I could make it any shorter 😉 
- naming convention. I used double underscore prefix. we could also create a `__bake` object and hook them onto that such as `__bake.filename` etc.

### CODE REVIEW

**Comment by david-zacharias** [comment]
Hey! And sorry for my late reply, have missed this PR.

I think it's a good idea to supply static information. 

 __timestamp__ is perfect. I also thought of `date`, `time`, etc., but if one wants another format it's possible to use a transform like `{{ __bake.timestamp | dateformat }}` where `dateformat` is defined by the developer.  

__filepath__ I was asking myself, is it the name of the source or of the destination file? Maybe we should supply both? Ideas for naming?

__filename__ If we supply it, it should definitely contain the name, not a path. If we want to keep the amount of variables low, filepath is enough, since one can also write a transform to extract dir or filename. On the other hand we should supply vars often used directly, to not force users to write to many transforms.

Regarding the naming I would go for `__bake.TYPE` to make clear where the data comes from and to prevent issues with other tasks that users may use for transforming files.

Regarding the implementation: I thought for a second about enabling developers to hand in own variables by suppling a hash or function via `bake.options`, but I think that's nothing for now. Maybe for bake 2.0.

**Comment by MathiasPaumgarten** [comment]
Yeah I agree, `__filename` should be enough. I guess I was kind of thinking node in the sense of `__dirname` and `__filename` but the one is just a subset of the other. 

I would not add any options for now. But that might be a nice extension for __2.0__. I think we are close to that one. Some cleanup work and adding this fix: #91 (Sadly that kind of "brakes" compilation on windows right now. Not actually but I could see some scenario where someone relies on the exact output [_including empty lines and such_] so it has to wait for a major update.)

So for now I will ad it to a `__bake` object. I would also like source and destination `__filename`. Not sure if this could get convoluted with some files having multiple destinations, maybe? Not sure. I can't think of an example right now but that somehow comes to mind. But I guess even if it did, that would be separate instances. 

So as for the variables:

```js
__bake.sourceFilename // with a synonym of __bake.filename
__bake.destFilename
__bake.timestamp
```

Anything else? Should we maybe kick out `sourceFilename` and just make it `filename`. I only added it for consistency.

**Comment by david-zacharias** [comment]
I think thats a good start, we can add further variables later on request. Although the names does not look nice with their camelCase-syntax, let's keep all for consistency. We can use `srcFilename` instead of `source` to align with Grunt's `src`, `dest` naming:


```js
__bake.srcFilename 
__bake.destFilename
__bake.filename // alias for srcFilename
__bake.timestamp
```

**Comment by MathiasPaumgarten** [comment]
@david-zacharias Wanna take a look? I think we can merge that for now. Pretty clean and simple. Barely changes the code base. Actually this PR adds more spaces for consistency than it adds actual new code 😄

**Comment by MathiasPaumgarten** [comment]
Also, tiny footnote, that has nothing today with this PR. Apparently, when you write something in single ticks (_`_) into a commit message, it gets taken out. The actual commit message of this was: 

__adds default variables to `__bake` object__

**Comment by david-zacharias** [comment]
Okay, now I know how the new GitHub Review feature works 👍 

P.S. Yes the single tick tricked me as well in another project... No day without surprises 🎉

**Comment by MathiasPaumgarten** [comment]
Coolio, will merge after I've added the README section.

**Comment by MathiasPaumgarten** [comment]
@nabto you will be pleased to hear that it's now available in `1.8.0`
